tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -359,6 +374,184 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
+ *       the opened batch is marked as the most recent batch (MRB).
+ *
+ *   (3) As soon as a tuple is requested from the heap, the following steps are performed:
+ *       (3a) If the heap is empty, we are done.
+ *       (3b) The top tuple from the heap is taken. It is checked if this tuple is from the
+ *            MRB. If this is the case, the next batch is opened, the top tuple is decompressed,
+ *            placed on the batch, and this batch is marked as MRB. This is repeated until the
+ *            top tuple from the heap is not from the MRB. This is done to ensure we have opened
+ *            all relevant batches (and one ahead) which might contain the most recent tuple
+ *            are opened and placed on the heap.
+ *
+ *            In the example above, the first three batches are opened because the first two
+ *            batches might contain tuples with a value of 0.
+ *       (3c) The top element from the heap is removed, the next tuple from the batch is
+ *            decompressed (if present) and placed on the heap.
+ *       (3d) The former top tuple of the heap is returned.
+ *
+ * This function checks if the compression 'order by' and the query 'order by' are
+ * compatible and the optimization can be used.
+ */
+static MergeBatchResult
+is_able_to_use_segment_merge_append(PlannerInfo *root, CompressionInfo *info, Chunk *chunk)
+{
+	int pk_index;
+	PathKey *pk;
+	Var *var;
+	Expr *expr;
+	char *column_name;
+	List *pathkeys = root->query_pathkeys;
+	FormData_hypertable_compression *ci;
+	ListCell *lc = list_head(pathkeys);
+	MergeBatchResult merge_result = SCAN_FORWARD;
+
+	/* Ensure that we have path keys and the chunk is ordered */
+	if (pathkeys == NIL || ts_chunk_is_unordered(chunk) || ts_chunk_is_partial(chunk))
+		return MERGE_NOT_POSSIBLE;
+
+	/*
+	 * Loop over the of pathkeys of the query. These pathkeys
+	 * need to exactly match the configured compress_orderby pathkeys.
+	 */
+	for (pk_index = 1; lc != NULL; lc = lnext_compat(pathkeys, lc), pk_index++)
+	{
+		pk = lfirst(lc);
+		expr = find_em_expr_for_rel(pk->pk_eclass, info->chunk_rel);
+
+		if (expr == NULL || !IsA(expr, Var))
+			return MERGE_NOT_POSSIBLE;
+
+		var = castNode(Var, expr);
+
+		if (var->varattno <= 0)
+			return MERGE_NOT_POSSIBLE;
+
+		column_name = get_attname(info->chunk_rte->relid, var->varattno, false);
+		ci = get_column_compressioninfo(info->hypertable_compression_info, column_name);
+
+		if (ci->orderby_column_index != pk_index)
+			return MERGE_NOT_POSSIBLE;
+
+		/* Test that ORDER BY and NULLS first/last do match */
+		if (ci->orderby_nullsfirst != pk->pk_nulls_first)
+			return MERGE_NOT_POSSIBLE;
+
+		/* Check order, if the order of the first column do not match, switch to backward scan */
+		if (ci->orderby_asc && pk->pk_strategy != BTLessStrategyNumber)
+		{
+			if (pk_index == 1)
+				merge_result = SCAN_BACKWARD;
+			else if (merge_result == SCAN_BACKWARD)
+				continue;
+			else
+				return MERGE_NOT_POSSIBLE;
+		}
+
```

Comment:
Would be good to fix the codecov complaints around this area.

Reply:
I refactored the logic and added more test cases. The missing lines are now covered by tests.